sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -350,28 +354,6 @@ export class BatchingReceiver extends MessageReceiver {
         reject(err);
       }, abortSignal);
 
-      // Use new message wait timer only in peekLock mode
-      if (this.receiveMode === ReceiveMode.peekLock) {
```

Comment:
Curious what your motivation is for having all modes use the new timeout after the first message is received now that we've gotten rid of debouncing. I'd think that if you were getting new messages quickly in receiveAndDelete mode, it'd be worth waiting for the maxWaitTime.

Reply:
This is just simplifying things (based on some of our internal chats).

I have another comment in another spot that explains why this is beneficial to both implementations (for different reasons).

For peekLock just getting the messages back earlier avoids the "messages are expiring before I see them"
For receiveAndDelete, getting the messages back earlier heads off a separate problem of "app crash resulted in me losing messages I'd never even seen".

So doing this kind of immediate-ish-return seems nice in both cases.

Reply:
For receiveAndDelete mode, we avoided returning early due to data loss reasons. See #5757
With the changes in this PR, I wonder if #5757 will be regressed.

Once credits are put on the link, rhea will start firing off events for every message received. The callback for each of these events will be queued up in the event loop. If we exit due to the new 1 second timer introduced in this PR, we lose all the messages for whom the callbacks were queued up in the event loop.

Draining credits will not help in this case as the messages have already reached rhea, they just have not reached the client.

We can consider having a buffer that will collect the messages that rhea has already received and pass them to the user in the subsequent receiveMessages() call.

Reply:
Related issue: #6349

Reply:
It seems like waiting longer is just a band-aid then - either way receiving messages at your time boundary is going to cause issues.

I'm curious if we simply await'd on drain if that would be enough. Carrying over state isn't bad either but I'd worry they wouldn't realize that they might have messages left over and we'd be back into the scenario of holding onto messages that the user has not seen.

Also, then we don't run into the "ask for 5 messages, get 'n' number of bonus messages we had cached".

Reply:
> It seems like waiting longer is just a band-aid then - either way receiving messages at your time boundary is going to cause issues.

Yes, but only if you are still receiving messages at the time boundary. Waiting for the entire `maxWaitTime` to complete gave a higher chance of getting all messages that are in flight and the ones that reached rhea, but not us than exiting after the first message. This is not a problem in other languages as they collect the in-flight messages in their buffer.

> I'm curious if we simply await'd on drain if that would be enough.

We do await on drain. But, drain has no affect on messages rhea has already received and fired off events for

> Carrying over state isn't bad either but I'd worry they wouldn't realize that they might have messages left over and we'd be back into the scenario of holding onto messages that the user has not seen.

Yes, and this is the state all other languages are at I believe. There is no way to 100% guarantee no data loss unless the user is asking for messages 1 by 1.

> Also, then we don't run into the "ask for 5 messages, get 'n' number of bonus messages we had cached".

Why would it be bonus? If `n` number of messages is cached, we empty those out first before adding credits on the line. I believe that is what we did in Event Hubs

Reply:
> > It seems like waiting longer is just a band-aid then - either way receiving messages at your time boundary is going to cause issues.
> 
> Yes, but only if you are still receiving messages at the time boundary. Waiting for the entire `maxWaitTime` to complete gave a higher chance of getting all messages that are in flight and the ones that reached rhea, but not us than exiting after the first message. This is not a problem in other languages as they collect the in-flight messages in their buffer.
> 
> > I'm curious if we simply await'd on drain if that would be enough.
> 
> We do await on drain. But, drain has no affect on messages rhea has already received and fired off events for

We don't do it if we think the number of credits on the receiver is zero. I'm curious if that's part of the issue that we see but without a repro it's hard for us to tell if we are doing the right thing here. 

We have a few issues related to potential message loss (in receiveAndDelete scenarios) that might be worth looking at.

> 
> > Carrying over state isn't bad either but I'd worry they wouldn't realize that they might have messages left over and we'd be back into the scenario of holding onto messages that the user has not seen.
> 
> Yes, and this is the state all other languages are at I believe. There is no way to 100% guarantee no data loss unless the user is asking for messages 1 by 1.

None of the other implementations are yet doing drain either so we might have something nicer we can offer here.

> 
> > Also, then we don't run into the "ask for 5 messages, get 'n' number of bonus messages we had cached".
> 
> Why would it be bonus? If `n` number of messages is cached, we empty those out first before adding credits on the line. I believe that is what we did in Event Hubs

Then we are potentially eternally holding onto messages (not the same ones, but it could be difficult to empty our internal buffer). 

Keeping in mind that a buffered approach will work but I think it's less favorable than just returning the messages without holding onto anything internally. So I'm just brainstorming about how that might work.

Reply:
> We don't do it if we think the number of credits on the receiver is zero. I'm curious if that's part of the issue that we see but without a repro it's hard for us to tell if we are doing the right thing here.

Ah.. interesting. So when you said "wait for drain", you meant, "keep collecting messages from rhea until drain is done" vs what we do now which is "stop collecting messages from rhea and then do drain". Yup, that change makes sense

Reply:
Yeah, we just need a good "close and flush" type mechanic.

Reply:
I need to look into this more - there appears to be some behavior around drain that makes it not work for the scenario I outlined above.

Reply:
I think we'll need a deeper look into why this doesn't work properly. For now going back to what we did before - not running the idle timeout check when receiveAndDelete is used as the lock mode.

Issues #6349 covers this - appears to still be an issue even with the latest dependencies.